src/blockchain/ain/jellyfish/utxo-provider.service.ts

```diff
@@ -0,0 +1,131 @@
+import { AddressUnspent } from '@defichain/whale-api-client/dist/api/address';
+import { Prevout } from '@defichain/jellyfish-transaction-builder';
+import BigNumber from 'bignumber.js';
+import { SmartBuffer } from 'smart-buffer';
+import { toOPCodes } from '@defichain/jellyfish-transaction';
+import { WhaleClient } from '../whale/whale-client';
+import { WhaleService } from '../whale/whale.service';
+import { Injectable } from '@nestjs/common';
+import { Config } from 'src/config/config';
+
+export interface UtxoInformation {
+  prevouts: Prevout[];
+  scriptHex: string;
+  total?: BigNumber;
+}
+
+@Injectable()
+export class UtxoProviderService {
+  private blockHeight = 0;
+  private unspent = new Map<string, AddressUnspent[]>();
+  private spent = new Map<string, NodeJS.Timeout>();
+
+  private whaleClient?: WhaleClient;
+
+  constructor(whaleService: WhaleService) {
+    whaleService.getClient().subscribe((client) => (this.whaleClient = client));
+  }
+
+  // --- QUEUED ENTRY POINTS --- //
+  async provideExactAmount(address: string, amount: BigNumber): Promise<UtxoInformation> {
+    const utxo = await this.retrieveUtxoAndPrevouts(address);
+    return this.markUsed(address, UtxoProviderService.provideExactAmount(address, utxo, amount));
+  }
+
+  async provideUntilAmount(address: string, amount: BigNumber): Promise<UtxoInformation> {
+    const utxo = await this.retrieveUtxoAndPrevouts(address);
+    return this.markUsed(address, UtxoProviderService.provideUntilAmount(utxo, amount));
+  }
+
+  // --- HELPER METHODS --- //
+  private markUsed(address: string, utxo: UtxoInformation): UtxoInformation {
+    utxo.prevouts.forEach((p) => {
+      const id = UtxoProviderService.idForPrevout(p);
+      this.spent.set(
+        id,
+        setTimeout(() => {
+          this.spent.delete(id);
+        }, Config.staking.timeout.utxo),
+      );
+    });
+    this.unspent.set(
+      address,
+      this.unspent.get(address)?.filter((u) => !utxo.prevouts.map((p) => p.txid).includes(u.vout.txid)),
```

Comment:
Is the txId enough for the check here? Or do we need the vout?

Reply:
I'm probably getting confused because we're comparing `UtxoInformation` and `AdressUnspent` here 🤔  would it be possible to do all the logic in either of them, instead of mixing it up? E.g. already store the `UtxoInformation` in the map and not a list of unspent. Or do the inverse and have `AdressUnspent` everywhere and only convert it to `UtxoInformation` at the end.

Reply:
Yes kinda agree, on the one hand we have the blockchain data which is represented as `AddressUnspent` and on the other side, we are already preparing for using it in raw-txs as `Prevout`. Basically they are the same thing and I have just reused what we already had and there parsing + counting was built in `Prevout` and our list of is built in `AddressUnspent`. I will invert it, and on marking it used, I will parse it from `AddressUnspent` to `Prevout`